### PR TITLE
graph: Add the option to show function call graph as mermaid diagram

### PR DESCRIPTION
**Pull request by wendyisdream**
This option shows the function call graph as a mermaid flowchart
for visualization.

- mermaid page : https://mermaid-js.github.io/mermaid/

The usage for this option is add --mermaid in graph command.

The result can be rendered in the browser so it is formatted as HTML.
The mermaid syntax itself can be rendered in other services
including a mermaid plugin like github or draw.io.

The function name cannot be used as a node id because some suffix
like 'end', 'self' is indicated to special usage in mermaid.
So 'id' is added to identify the node to 'struct uftrace_graph_node'.

Signed-off-by: Jia Ha <wendyisdream1004@gmail.com>

### CODE REVIEW

**Comment by honggyukim** [comment]
It looks nice. Thanks very much for doing this work!
image

I have two comments on this work.
1. I understand this work is more related to `graph` than `dump`. If this is implemented in `graph` command, then `--graphviz` is also better to be moved in `graph` command. Moreover, I would like to hear how @namhyung thinks about this.

2. `uftrace graph` has a partial graph mode. The usage of `graph` is as follows.
```
    uftrace graph [options] [FUNCTION]
```
Let's say there is an example program and it can be compiled and recorded as follows.
```
$ gcc -pg -o t-racycount tests/s-racycount.c -pthread

$ uftrace record t-racycount
```
The `uftrace graph` shows its call graph from the top.
```
$ uftrace graph
# Function Call Graph for 'a.out' (session: 240a6ef673ec8bcf)
========== FUNCTION CALL GRAPH ==========
# TOTAL TIME   FUNCTION
    1.864 ms : (1) a.out
    2.130 us :  +-(1) __monstartup
             :  |
    1.163 us :  +-(1) __cxa_atexit
             :  |
    1.134 ms :  +-(1) main
    1.273 us :  |  +-(1) pthread_barrier_init
             :  |  |
  358.416 us :  |  +-(2) pthread_create
             :  |  |
  639.737 us :  |  +-(1) racy_count
  306.615 us :  |  | (1) pthread_barrier_wait
  294.710 us :  |  | (1) linux:schedule
             :  |  |
  128.560 us :  |  +-(2) pthread_join
  115.741 us :  |    (1) linux:schedule
             :  |
  726.979 us :  +-(2) thread_fn
  720.392 us :    (2) racy_count
   94.733 us :    (1) linux:schedule
```
If `FUNCTION` is given as `racy_count`, it builds a call graph from the `FUNCTION` and it also collects all the backtrace that reaches to the `FUNCTION`.
```
$ uftrace graph racy_count
# Function Call Graph for 'racy_count' (session: 240a6ef673ec8bcf)
=============== BACKTRACE ===============
 backtrace #0: hit 2, time 720.392 us
   [0] thread_fn (0x4009a5)
   [1] racy_count (0x40095f)

 backtrace #1: hit 1, time 639.737 us
   [0] main (0x4009c9)
   [1] racy_count (0x40095f)

========== FUNCTION CALL GRAPH ==========
# TOTAL TIME   FUNCTION
    1.360 ms : (3) racy_count
  306.615 us :  +-(1) pthread_barrier_wait
  294.710 us :  | (1) linux:schedule
             :  |
   94.733 us :  +-(1) linux:schedule
```
This case has to be covered in `--mermaid` option if it's used in `graph` command.

**Comment by namhyung** [comment]
I think it's better to have it in dump command as it's for external tools.

**Comment by wendyisdream** [comment]
@honggyukim, @namhyung Thank you for the review. 
Okay, I will move this option to dump command part. 
Oh! uftrace have the partial graph mode!
I will try to include it in this option.

**Comment by wendyisdream** [comment]
I updated the code to move the mermaid option to the dump command.
I tried to add this option following the existing dump option's format to remove this PR's other effect.
I added the code to be shown as a single root to graphviz and mermaid which is raised in #1499.

In dump command, adding a partial graph mode and handling multiple sessions, and refactoring duplicate code between commands is needed but I think it will be better to handle in each separate PR.

**Comment by namhyung** [comment]
In dump command, we don't support the partial graph.  But we can use filters to limit the output.

**Comment by honggyukim** [comment]
> I added the code to be shown as a single root to graphviz and mermaid which is raised in https://github.com/namhyung/uftrace/issues/1499.

Thanks for the update, but it's not a good idea to handle two different tasks in a single commit. Please separate them in different patches.

**Comment by wendyisdream** [comment]
I deleted the update in graphviz to separate from #1499. :)
Thank you for the review!

File: cmds/dump.c [hunk 1]
```diff
@@ -1244,6 +1248,126 @@ static void dump_graphviz_footer(struct uftrace_dump_ops *ops, struct uftrace_da
...
+	if (opts->caller)
+		pr_out(" /caller: %s", opts->caller);
+	if (opts->hide)
+		pr_out(" /hide: %s", opts->hide);
+	if (opts->tid)
+		pr_out(" /tid: %s", opts->tid);
+	if (opts->depth != OPT_DEPTH_DEFAULT)
+		pr_out(" /depth: %d", opts->depth);
+	if (opts->threshold) {
+		pr_out(" /threshold:");
```
Comment (namhyung) [inline cmds/dump.c:1345]: Please print it as "time-filter" as it's the name of the option.

File: utils/graph.c [hunk 2]
```diff
@@ -87,6 +88,7 @@ static int add_graph_entry(struct uftrace_task_graph *tg, char *name, size_t nod
 
 		node = xzalloc(node_size);
 
+		node->id = next_id++;
```
Comment (namhyung) [inline utils/graph.c:91]: Although this is a small change, I think it's better to split it as a separate commit because it's a general change.  Please add a comment to `graph_add_node()` function that it's not thread-safe due to the static id.

**Comment by wendyisdream** [comment]
Thank you for the review!
I modified it as commented and separated this PR into three commits.

b518c3c2 dump: Add a single root node as the executed file in graphviz
9bc8bc38 dump: Add the option to show function call graph as mermaid diagram
493ea934 graph: Add id to uftrace_graph_node to identify graph node

File: cmds/dump.c [hunk 3]
```diff
@@ -1250,6 +1254,126 @@ static void dump_graphviz_footer(struct uftrace_dump_ops *ops, struct uftrace_da
...
+static void dump_mermaid_header(struct uftrace_dump_ops *ops, struct uftrace_data *handle,
+				struct uftrace_opts *opts)
+{
+	graph_init_callbacks(NULL, NULL, NULL, ops);
+	mermaid_graph.root.name = basename(handle->info.exename);
+
+	pr_out("<html>\n");
+	pr_out("  <body>\n");
+	pr_out("    <script src=\"https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js\"></script>\n");
+	pr_out("    <script>\n");
```
Comment (honggyukim) [inline cmds/dump.c:1312]: We don't need to add whitespaces for indentation for html tags. Please remove them here and rest of the other places.

I'm okay with having spaces for javascript code.

File: cmds/dump.c [hunk 4]
```diff
@@ -1250,6 +1254,126 @@ static void dump_graphviz_footer(struct uftrace_dump_ops *ops, struct uftrace_da
...
+}
+static void dump_mermaid_footer(struct uftrace_dump_ops *ops, struct uftrace_data *handle,
+				struct uftrace_opts *opts)
+{
+	/* skip empty graph */
+	if (mermaid_graph.root.time == 0 && mermaid_graph.root.nr_edges == 0)
+		return;
+
+	if (mermaid_graph.root.time || mermaid_graph.root.nr_edges) {
+		pr_out("    <h2>Function Call Graph for <span style=\"color:blue\">%s</span>",
```
Comment (honggyukim) [inline cmds/dump.c:1366]: Please remove spaces before `<h2`>.

**Review by honggyukim — COMMENTED** [review]
Thanks for the update. It looks good in general, but I have a few more requests.

File: utils/graph.c (hunk 2 above)
Comment (wendyisdream) [reply utils/graph.c:91]: Thank you for the review!
I added the comment for it to line 210 because graph_add_node is a public interface for others.
Do you think it is better to move it here?
Except it, I updated the code.

Comment (wendyisdream) [reply utils/graph.c:91]: > > Please add a comment to graph_add_node() function that it's not thread-safe due to the static id.
> 
> Please address @namhyung's comment above.

Thank you for the review!
I added the comment for utils/graph.c to line 210 because graph_add_node is a public interface for others.
Do you think it is better to move it here?
Except for it, I updated the code.

**Comment by honggyukim** [comment]
Thanks for the update. It's almost done, but could you add a test in `tests/t272_dump_mermaid`. You can make it similar as `tests/t101_dump_chrome.py`.

**Comment by wendyisdream** [comment]
Testcase 272 is updated to 5aad7f4 commit. Thank you for informing of it!

File: cmds/dump.c [hunk 5]
```diff
@@ -1250,6 +1254,104 @@ static void dump_graphviz_footer(struct uftrace_dump_ops *ops, struct uftrace_da
...
+	pr_out("</script>\n");
+}
+
+static void print_graph_node_mermaid(struct uftrace_graph *graph, struct uftrace_graph_node *node)
+{
+	char *symname = node->name;
+	struct uftrace_graph_node *child;
+
+	if (node->addr == EVENT_ID_PERF_SCHED_IN)
+		symname = "linux:schedule";
```
Comment (namhyung) [inline cmds/dump.c:1328]: I don't think we need this.  It should be handled by `graph_add_node()` already.

**Review by namhyung — COMMENTED** [review]
Mostly looks good.  A nit below..

File: cmds/dump.c (hunk 5 above)
Comment (wendyisdream) [reply cmds/dump.c]: I applied the code to delete this :)

**Review by namhyung — COMMENTED** [review]
Looks good.  Could you please add an example output (and the comment line) in the second commit message even if it's in the test case?  I think you can just copy the same output of the test case.  Thanks!

**Comment by wendyisdream** [comment]
I added the runtest.py result to the commit message. :)

**Comment by wendyisdream** [comment]
I deleted unnecessary brackets in a python pattern expression.

**Review by namhyung — APPROVED** [review]
LGTM
